### PR TITLE
Recovery indexer.

### PR DESCRIPTION
**Pull request by mansu**
This diff adds the logic for the recovery indexer. 

A recovery indexer performs a recovery task and uploads a chunk to S3. To support this, we added a new chunk implementation called `RecoveryChunkImpl` that provides a ReadWriteChunk api, but doesn't publish live snapshots for the query service to query. 

A `RecoveryChunkManager` implements a chunk manager for the recovery task. 

I have also added a new parallel kafka consumer, that can ingest chunk data in parallel into the chunk manager as oppsed to the single threaded ingestion in the `IndexingChunkManager`.  

Updated the RecoveryService to use the new parallel kafka consumer, recovery chunk manager and recovery chunk implementation to implement recovery functionality.

### CODE REVIEW

File: kaldb/src/test/java/com/slack/kaldb/server/KaldbTest.java [hunk 1]
```diff
@@ -134,12 +133,12 @@ private KaldbConfigs.KaldbConfig makeKaldbConfig(
...
     return KaldbConfigUtil.makeKaldbConfig(
         "localhost:" + kafkaServer.getBroker().getKafkaPort().get(),
-        indexPort,
+        port,
         kafkaTopic,
         kafkaPartition,
         clientName,
         TEST_S3_BUCKET,
-        queryPort,
+        port + 1,
```
Comment (bryanlb) [inline kaldb/src/test/java/com/slack/kaldb/server/KaldbTest.java:141]: Looks like this was inadvertently reverted in a rebase to the previous logic. Should continue to be explicitly defined.

**Review by bryanlb — CHANGES_REQUESTED** [review]

File: kaldb/src/main/java/com/slack/kaldb/chunkManager/NeverRolloverChunkStrategyImpl.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.slack.kaldb.chunkManager;
+
+public class NeverRolloverChunkStrategyImpl implements ChunkRollOverStrategy {
```
[the remark this answers is not in the document]
Comment (mansu) [reply kaldb/src/main/java/com/slack/kaldb/chunkManager/NeverRolloverChunkStrategyImpl.java:3]: Done.

File: kaldb/src/main/java/com/slack/kaldb/writer/kafka/KafkaOffsetBoundsExceededException.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.slack.kaldb.writer.kafka;
+
+/**
+ * KafkaMaxOffsetExceededException is thrown when the consumer is trying to index past the max
+ * allowed offset for a Kafka topic partition.
+ */
+public class KafkaOffsetBoundsExceededException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (mansu) [reply kaldb/src/main/java/com/slack/kaldb/writer/kafka/KafkaOffsetBoundsExceededException.java:7]: Let's leave it in for now. I think we need to catch this exception in the recovery service if needed.

File: kaldb/src/main/java/com/slack/kaldb/writer/kafka/KaldbKafkaConsumer.java [hunk 4]
```diff
@@ -211,6 +213,90 @@ public void consumeMessages(long kafkaPollTimeoutMs) throws IOException {
     }
   }
 
+  private static class BlockingArrayBlockingQueue<E> extends ArrayBlockingQueue<E> {
```
[the remark this answers is not in the document]
Comment (mansu) [reply kaldb/src/main/java/com/slack/kaldb/writer/kafka/KaldbKafkaConsumer.java:223]: Added.

File: kaldb/src/main/java/com/slack/kaldb/recovery/RecoveryService.java [hunk 5]
```diff
@@ -31,35 +43,47 @@
...
-      "recovery_node_received_assignment";
-  public static final String RECOVERY_NODE_COMPLETED_ASSIGNMENT =
-      "recovery_node_completed_assignment";
-  public static final String RECOVERY_NODE_FAILED_ASSIGNMENT = "recovery_node_failed_assignment";
-  protected final Counter recoveryNodeReceivedAssignment;
-  protected final Counter recoveryNodeCompletedAssignment;
-  protected final Counter recoveryNodeFailedAssignment;
+  public static final String RECOVERY_NODE_ASSIGNMENT_RECEIVED =
+      "recovery_node_assignment_received";
+  public static final String RECOVERY_NODE_ASSIGNMENT_SUCCESS = "recovery_node_assignment_success";
```
[the remark this answers is not in the document]
Comment (mansu) [reply kaldb/src/main/java/com/slack/kaldb/recovery/RecoveryService.java:69]: I couldn't find a way to access the time series in unit tests using tags. So, emitting them as a separate time series for now.

File: kaldb/src/main/java/com/slack/kaldb/writer/kafka/KafkaOffsetBoundsExceededException.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/writer/kafka/KafkaOffsetBoundsExceededException.java:7]: We don't seem to actually catch this custom exception - if we have code that handles this scenario separately (ie, catching it to do something with it) then we can keep it in. If we're just adding a new exception type and not using it let's leave it out.

File: kaldb/src/main/java/com/slack/kaldb/server/RecoveryTaskCreator.java [hunk 6]
```diff
@@ -103,8 +103,8 @@ public static long getHighestDurableOffsetForPartition(
     return Math.max(maxRecoveryOffset, maxSnapshotOffset);
   }
 
-  private static String getRecoveryTaskName(long creationTimeEpochMs, String partitionId) {
-    return "recoveryTask_" + partitionId + "_" + creationTimeEpochMs;
+  private static String getRecoveryTaskName(String partitionId) {
+    return "recoveryTask_" + partitionId + "_" + Instant.now().getNano();
```
Comment (bryanlb) [inline kaldb/src/main/java/com/slack/kaldb/server/RecoveryTaskCreator.java:107]: Is there a reason for using a nano here? We can't have contention trying to create a recovery task for the same partition ID, and this makes the znode path much longer than is needed.

File: kaldb/src/test/java/com/slack/kaldb/testlib/KaldbConfigUtil.java [hunk 7]
```diff
@@ -74,10 +74,20 @@ public static KaldbConfigs.KaldbConfig makeKaldbConfig(
                     .build())
             .build();
 
+    KaldbConfigs.RecoveryConfig recoveryConfig =
+        KaldbConfigs.RecoveryConfig.newBuilder()
+            .setServerConfig(
+                KaldbConfigs.ServerConfig.newBuilder()
+                    .setServerPort(indexerPort + 3)
```
Comment (bryanlb) [inline kaldb/src/test/java/com/slack/kaldb/testlib/KaldbConfigUtil.java:81]: This needs to not use a magic number per https://github.com/slackhq/kaldb/pull/249

**Review by bryanlb — CHANGES_REQUESTED** [review]

File: kaldb/src/main/java/com/slack/kaldb/writer/kafka/KafkaOffsetBoundsExceededException.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mansu) [reply kaldb/src/main/java/com/slack/kaldb/writer/kafka/KafkaOffsetBoundsExceededException.java]: Ok dropped it for now.

File: kaldb/src/main/java/com/slack/kaldb/server/RecoveryTaskCreator.java (hunk 6 above)
Comment (mansu) [reply kaldb/src/main/java/com/slack/kaldb/server/RecoveryTaskCreator.java:107]: I used millis initially and ended up with some race conditions in unit tests so I switched to nano.  It's about 13 chars as oppsed to 10 so it was not too bad in terms of string length. So kept it as is.

File: kaldb/src/test/java/com/slack/kaldb/testlib/KaldbConfigUtil.java (hunk 7 above)
Comment (mansu) [reply kaldb/src/test/java/com/slack/kaldb/testlib/KaldbConfigUtil.java:81]: Done.

**Review by bryanlb — APPROVED** [review]

**Pull request merged**